### PR TITLE
Convert CONTRIBUTING.md into a Sway Application Book

### LINKED ISSUE #141: Convert CONTRIBUTING.md into a Sway Application Book
### Motivation

In the upcoming changes to the doc (#130) there are snippets of code embedded in the file and the file is large.
It will be a good idea to split up the file into sections, and remove the code (put it into individual files), so that it is easier to navigate and generally add to the Book.
This has the additional benefit of allowing us to validate the code and including it as a snippet directly in the book.

### PR DESCRIPTION
**Pull request by Braqzen**
## Type of change

- Documentation

## Changes

The following changes have been made:

- Converted contributing markdown doc into `mdbook`
- Updated, added, and removed, various sections from the previous iteration
- Added new file to CI for deployment whenever a new version hits master (hopefully it works)
- Removed a line in the current CI for apps which was unused
- Deleted the old contributing doc

## Notes

- Need to update link in README to point to live book once that is set up (after this PR goes in the github page can be set up)
- Other projects will also need to have that link updated

## Related Issues

Closes #141

### CODE REVIEW

File: .docs/contributing-book/src/documentation/project-quality/testing.md
Comment (nfurfaro) [inline .docs/contributing-book/src/documentation/project-quality/testing.md:49]: "All assertions & requirements should be tested..."

File: .docs/contributing-book/src/documentation/project-quality/testing.md
Comment (nfurfaro) [inline .docs/contributing-book/src/documentation/project-quality/testing.md:47]: As per my other comment about code structuring, I would prefer to add tests for assertions/requirements in the same order as the assertion/requirements being tested occur in the code, as opposed to imposing a hard requirement of alphabetical ordering on them.

File: .docs/contributing-book/src/documentation/pull-requests/commit.md
Comment (nfurfaro) [inline .docs/contributing-book/src/documentation/pull-requests/commit.md:43]: I've always gone with "Fix" over "Fixed" in this example. That is, use the imperative voice (which seems to be the generally accepted best practice. 
Both links provided below, as well as this [one](https://cbea.ms/git-commit/) follow this practice.

**Review by nfurfaro — COMMENTED** [review]
I left a few comments, but looking really good!

File: .docs/contributing-book/src/code/bad_documentation/tests/harness.rs
Comment (Braqzen) [reply .docs/contributing-book/src/code/bad_documentation/tests/harness.rs:40]: I have considered removing all tests since none are currently implemented however a counter point was to leave the harness in because otherwise it would need to be hand coded from scratch if it becomes relevant. Of course, we could simply bump the version and automatically generate a new one for us. I have chosen to leave this in for now in case someone wants to easily extract some test code and because it doesn't cost us anything to have it there especially since it's "hidden" from the public

File: .docs/contributing-book/src/documentation/pull-requests/commit.md
Comment (Braqzen) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:43]: The point isn't about "Fix" vs "Fixed" it's about those semantically devoid commit messages being bad. There needs to be more content to paint a context. That being said... don't look at my commit log :)

File: .docs/contributing-book/src/documentation/project-quality/testing.md
Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/testing.md:47]: Turns out there is a point below which mentions this. I have moved it as a sub-point here

File: .docs/contributing-book/src/documentation/project-quality/code-structure.md
Comment (simonr0204) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: I agree that alphabetical ordering is best _within_ code structures (struct/enum fields, function params, etc.) but I think we should leave the order of those structures within the file to the discretion of the contributor, or at least soften the "must" requirement to a "should" (i.e. a good default to go with unless there's a compelling reason not to).

For example, it often makses sense to group functions by permission level - e.g. permissioned admin functions.

File: .docs/contributing-book/src/documentation/pull-requests/commit.md
Comment (simonr0204) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:43]: I'm gonna need an exception to this rule for `"format"` 😅

File: .docs/contributing-book/src/documentation/project-quality/code-structure.md
Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: Yes you can group things together, and the alphabetical thing annoys me at times too, nevertheless it does not counter the argument of increased cognitive load because you potentially have to familiarise yourself with the code each time you see it if you've not seen it in a while or if things have grown over time. If a new structure is proposed that is closer to grouping and is in similar cognitive load i.e. "learn this rule instead of alphabetical rule here" then I'll be fine with that otherwise it allows devs to throw things where ever any one of them sees fit and I rather not spend my attention trying to figure out how to traverse any new codebase or how things have changed.

Comment (simonr0204) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: I'm not sure I follow that alphabetical order necessarily reduces cognitive load. It's a handy convention, but if I encounter code I'm not familiar with, it's rare for me to think "right, where can I find the functions that begin with 'J' " 🙃 

Happy to go with the consensus, once reached. But I do think this should be mulled over a bit, since it sounds like not all developers (from a sample size of 3 😁) approach this this same way.

I'm also keen to find a happy medium that respects the reviewer's time but also lets the developer work without feeling too constrained.

Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: My last comment was a nudge for a proposition of an alternative. Atm there are 2 devs that at least follow, but I don't know if they agree with, the convention. If an immediate proposition is not going to be made then there is no reason for this PR to sit here when it can be approved and updated in the future if a proposition comes in.

Comment (nfurfaro) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: Solidity has an established convention in terms of style & location of elements in a file.
It isn't all applicable here, as we have:
1.) `forc fmt` to enforce universal formatting
2.) good library support to avoid the need to dump everything needed by a contract into 1 file.

That said, establishing a convention about function order based on purity would be in line with what solidity does.
We could extend that to cover cases where the dev wants to use a single file structure, i.e: a small contract with 1 error, 1 event, 1 private(util) function and 1 data structure should not be forced to uses a 5-file project structure in my opinion.
For a more complex project, I really like the multi-module setup though, so I think allowing leeway for use of judgement makes sense.

Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: Can you provide more information regarding purity i.e. where do read only functions go, where do write only functions go, where do read/write functions go, what about the ordering of those functions within their own block of functions?
I think that structure is acceptable once a concrete structure is put into place.

Something that I'm not a fan of is grouping by concept e.g. "admin things here" because now I have to learn about each project to know where things are and there are a lot of stipulations and exceptions that would need to be listed as opposed to the 3 cases for purity in Sway.

I'm indifferent to content being placed in 1 file as long as it's not an actual "project" / "application" because that isn't an application imo. There's not enough code to warrant that as an app therefore I don't think that applies to the apps repo nor should that content be placed there.
There might be some outliers in early development e.g. the OTC swap predicate may not need additional functionality and can all fit in 1 file _at the moment_.

I'm fine with people structuring things in a consistent and logical way in a larger project. So far I have advocated for a `data_structures.sw` _file_ but that may be inappropriate for some projects and thus a directory of individual sway files might apply (as an example).

So, if information can be provided about the exact rules then I can come up with some text and update the PR.

Comment (simonr0204) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: It sounds to me like there is a strong case for _not_ having stringent rules for ordering. I definitely acknowledge that there's a trade-off between cognitive overhead for the developer and the reviewer, but i'm hearing a preference amongst _most_ contributors for a little more freedom in this regard.

I would be in favor of dropping this requirement until there is some consensus. If there's no buy-in from contributors at large, it won't stick anyway.

Comment (nfurfaro) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: @Braqzen https://docs.soliditylang.org/en/v0.8.17/style-guide.html#order-of-functions

Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: The only relevant thing I see comes to the following ordering

1. `write`
2. `read`
3. `pure`

I think all `writes` require the `read` annotation as well so 

1. `write & read`
2. `read`
3. `pure`

for functions inside a contract. I don't really see the point of alphabetically ordering these. You could but I'm uncertain if that would be more useful than the amount of work needed to do this.

So, is the following acceptable:

1. variables / fields in data structures should be alphabetical
2. group data structures by module and alphabetically order the content there
3. when testing order the functions in the order of the requirements being tested
4. abi functions should be ordered as stated above (in terms of purity)
5. imports and dependencies should be alphabetical
6. document the interface in alphabetical order but the parameters themselves can be in any order in a function

If I've missed anything out then I might edit or add another message

File: .docs/contributing-book/src/documentation/pull-requests/commit.md
Comment (nfurfaro) [inline .docs/contributing-book/src/documentation/pull-requests/commit.md:39]:
```suggestion
  - Fix a conditional check in the deposit function which did not fail under condition XYZ
```

File: .docs/contributing-book/src/documentation/project-quality/code-structure.md
Comment (nfurfaro) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: I think the above looks pretty good to me. 
For point # 6, I would probably prefer to document the interface params in the order they are declared in the function signature as opposed to alphabetical, but it's not a hill I need to die on. Happy to go with the majority here.

Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: Awaiting confirmations from @simonr0204 @matt-user @bitzoic @SwayStar123 @supiket

File: .docs/contributing-book/src/documentation/pull-requests/commit.md
Comment (Braqzen) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:39]: I'm pretty sure people do not say "I fix a conditional check" and instead they say "I have fixed a conditional check". The example above omits the first two words.
I also do not subscribe to the idea of using "Fix, Chore, whatever" in the commit message. People are free to do that but that's only useful when resolving bugs that have sprung up and the log is being unwound. Labels and issues exist for this purpose otherwise.

Comment (nfurfaro) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:39]: Whether or not we use so-called "semantic commits" is another issue.

The change I suggested here is to use the imperative mood in the commit message ("fix" vs "fixed"), as recommended by pretty much all the guides to writing good commit messages (including the 2 linked in this book section.

Comment (nfurfaro) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:39]: Git itself uses this convention. After a merge or a revert, the corresponding commit messages are "Merge ..." or "Revert ..." , not "Merged ..." or "Reverted ..."

File: .docs/contributing-book/src/documentation/project-quality/code-structure.md
Comment (SwayStar123) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: I think "getter" functions should be separated / put at the top/bottom of the abi impl block, it doesnt make sense to mix them into the functions that actually do stuff

File: .docs/contributing-book/src/documentation/pull-requests/commit.md
Comment (Braqzen) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:39]: I have read most of the article and my take away is to write concise messages.

I 100% disagree with the imperative tone. It's incorrect and I will not follow it. Who exactly are you commanding when you write such a message? The reader, yourself, the commit? It doesn't make sense.
The process is that you do some work and _describe_ the work done. You do not write a command as if that work now needs to be done because you've literally just done it and committed it. It's a bad standard and it's grammatically incorrect.

File: .docs/contributing-book/src/documentation/project-quality/code-structure.md
Comment (simonr0204) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: > I think the above looks pretty good to me. For point # 6, I would probably prefer to document the interface params in the order they are declared in the function signature as opposed to alphabetical, but it's not a hill I need to die on. Happy to go with the majority here.

wait, the order of function _params_ in the abi and in the implementation need to match, right? The function signatures have to be identical.  i.e. `myfunc(a: u64, b: Address)` != `myfunc(b: Address, a: u64)` ?

Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: It does make sense. It depends on which perspective you impose on yourself. For some reason your choice is a separation between getters and non-getters rather than some other rule. This rule is "okay" but it allows a looser format than via purity.

Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: > > I think the above looks pretty good to me. For point # 6, I would probably prefer to document the interface params in the order they are declared in the function signature as opposed to alphabetical, but it's not a hill I need to die on. Happy to go with the majority here.
> 
> wait, the order of function _params_ in the abi and in the implementation need to match, right? The function signatures have to be identical. i.e. `myfunc(a: u64, b: Address)` != `myfunc(b: Address, a: u64)` ?

Yes, they must match however the documentation (the `///`) is what I was referring to. I want that to be alphabetical but I can live with the params being out of order although I would prefer them to be ordered. The `ABI` and implementation signatures must match though.

Comment (simonr0204) [reply .docs/contributing-book/src/documentation/project-quality/code-structure.md:9]: ohh I see. Yeah that's fine.

#shipit

File: .docs/contributing-book/src/documentation/pull-requests/commit.md
Comment (nfurfaro) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:39]: image

Comment (nfurfaro) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:39]: I think the key here is to think in terms of:
"When merged into master, this patch will ... (i.e: Fix xyz)

**Review by simonr0204 — APPROVED** [review]

File: .docs/contributing-book/src/documentation/pull-requests/commit.md
Comment (simonr0204) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:39]: I wouldn't get too bogged down on this. It's good to have guidelines but ultimately git commit messages are essentially immutable (I think you can `amend` them but it's basically a patch commit on top of the old one).

I think "concise" is a good goal, but taking a look at the average commit message in this repo, we have a ways to go before we start being concerned about tense/mood of the comment.

Unlike guidelines relating to code quality, commit messages should just be on a best-effot basis.

Comment (Braqzen) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:39]: They actually think that prepending some text in your mind is acceptable. How about I completely omit a message and they read my entire mind to figure out the commit? It's absolutely absurd that they think it's acceptable and just because github follows it doesn't mean it's good. Look at who has the most influence in society as an example. Is it the correct groups such as philosophers, teachers etc. or dumb celebrities that start mindless trends which everyone follows. The argument that "they do it" is subpar at best. The standard is meant to be judged on its own quality and not adoption.

That being said, I don't care about what someone decides to put in the message as long as it is concise and accurate. Although if someone starts adding emojis or something then that will annoy me.

I need to update the section with the points stated above since it looks like enough eyes have generally agreed. I will likely do that on Sunday so I will mark this back as a draft for now and rerequest a review once the changes are made.

Comment (simonr0204) [reply .docs/contributing-book/src/documentation/pull-requests/commit.md:39]: > as long as it is concise and accurate.

👍

**Review by bitzoic — CHANGES_REQUESTED** [review]
A lot of little things here and there but take my suggestions as you will. I think the biggest thing is some of the sentences could be broken down into smaller chunks or even multiple sentences so that they are more digestible and easier to read. 

The CONTRIBUTING.md file has been removed so the README.md link is now broken, this should either be removed or revised. The README.md should also link to this book so users can find it.

This has also been discussed at ethBerlin but it may be worthwhile to provide a link to other books such as the Sway Book, SDK book, and FuelUp book. I was helping out a few people and they were struggling to find the other books for more information. Providing a link would give easy access.

File: .docs/contributing-book/src/code/bad_documentation/tests/harness.rs
Comment (Braqzen) [reply .docs/contributing-book/src/code/bad_documentation/tests/harness.rs:36]: https://github.com/FuelLabs/sway-applications/pull/218#discussion_r975319259

File: .docs/contributing-book/src/documentation/project-quality/documentation/code/api.md
Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/documentation/code/api.md:1]: API is a common term and the ABI is basically the same thing but specialized to blockchain afaik. In order to make it more accessible I have opted for API instead of ABI. I'm open to further comments regarding this.

File: .docs/contributing-book/src/documentation/project-quality/documentation/code/comments.md
Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/documentation/code/comments.md:15]: I haven't looked into anchors so I don't know if they can be auto generated. I assume that you must write all of that into the file yourself therefore I have opted for this to keep the code cleaner and the docs.

File: .docs/contributing-book/src/documentation/project-quality/documentation/read-me.md
Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/documentation/read-me.md:13]: Nope, I meant the platform such as operating system etc.

**Comment by Braqzen** [comment]
I'll make a few minor adjustments tomorrow @bitzoic for the comments that were not commit suggestions. That being said, I have accepted basically all of them. I will point out that most of the commas are incorrect afaik since a sentence should not insert commas just because that is how someone would read it. A comma is meant to separate content rather than be a temporary pause to aid reading (which btw may differ based on who reads it). Nevertheless, if it makes it easier for people to read then I'll take that over correctness.

File: .docs/contributing-book/src/documentation/project-quality/documentation/code/comments.md
Comment (bitzoic) [reply .docs/contributing-book/src/documentation/project-quality/documentation/code/comments.md:15]: [Here](https://github.com/FuelLabs/sway/blob/ac745eaca9b4d42e6e216028f429453f8c3f467b/docs/src/basics/blockchain_types.md?plain=1#L54) is an example of where I used anchors in the Sway book and the corresponding [anchor](https://github.com/FuelLabs/sway/blob/ac745eaca9b4d42e6e216028f429453f8c3f467b/sway-lib-std/src/identity.sw#L7)

Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/documentation/code/comments.md:15]: Right, I have seen them before however I am unconvinced that it's worth the effort. The argument I see to use them is that if the file changes, changes to the code or formatting, then the book documentation will need to be update rather than just the code.
Again, I'm unsure if it's worth the effort of putting these throughout the files.

I'm open to it and now that I've realized that I cannot think of a counter to my argument I will probably do this... man, autism is meant to make this easier and not trigger my perfectionism...

**Comment by Braqzen** [comment]
Version `0.24.4` formatter is disfunctional in terms of an empty struct containing a comment. Hopefully bumping to the latest release (in another PR) will fix this.

**Review by bitzoic — DISMISSED** [review]
The CONTRIBUTING.md link in the src README.md file still doesn't point to anything and returns a 404. Can it point to this new book?

Besides that, looks good to me!

**Review by simonr0204 — APPROVED** [review]
Approving, as ever, with the caveat that these are guidelines, not law. Certainly, one must know the rules before being allowed to break them, but the latter will happen and we must remain pragmatic when it does.

This should be a living document that evolves as our understanding of what is best practice does.

**Review by matt-user — APPROVED** [review]

File: .docs/contributing-book/src/documentation/project-quality/documentation/code/comments.md
Comment (Braqzen) [reply .docs/contributing-book/src/documentation/project-quality/documentation/code/comments.md:15]: Correct. It's not markdown and that is why it's suggested to run the book (it's in the readme). Alternatively you can view the raw code.

**Review by matt-user — APPROVED** [review]

**Review by supiket — APPROVED** [review]
Not sure if it's worth mentioning in the book but I would like to point out that [configuration-time constants](https://fuellabs.github.io/sway/v0.24.4/basics/variables.html#configuration-time-constants) are snake_case, rather than SCREAMING_SNAKE_CASE.

**Comment by Braqzen** [comment]
> Not sure if it's worth mentioning in the book but I would like to point out that [configuration-time constants](https://fuellabs.github.io/sway/v0.24.4/basics/variables.html#configuration-time-constants) are snake_case, rather than SCREAMING_SNAKE_CASE.

I see. The book is incorrect then since they are constants. I'll keep that in mind when I'm writing the new book.

**Pull request merged**
